packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -87,11 +88,21 @@ export function buildImportAnalysisPlugin(config: ResolvedConfig): Plugin {
   const ssr = !!config.build.ssr
   const insertPreload = !(ssr || !!config.build.lib)
   const isWorker = config.isWorker
+  const rollupOptionsOutput = config.build?.rollupOptions?.output
+  const hasCustomAssetFileNames =
+    rollupOptionsOutput &&
+    !Array.isArray(rollupOptionsOutput) &&
+    !!rollupOptionsOutput.assetFileNames
 
   const scriptRel = config.build.polyfillModulePreload
     ? `'modulepreload'`
     : `(${detectScriptRel.toString()})()`
-  const preloadCode = `const scriptRel = ${scriptRel};const seen = {};const base = '${preloadBaseMarker}';export const ${preloadMethod} = ${preload.toString()}`
+  const assetsURL = hasCustomAssetFileNames
+    ? config.base
+    : isRelativeBase(config.base)
+    ? `new URL('./',import.meta.url).pathname`
```

Comment:
Just a thought that maybe for Vite 3, we should preserve `import.meta.url` when Vite transform `new URL(...)` too. Currently it's lowering it [by default](https://github.com/vitejs/vite/blob/83d32d965e3abd78fe390e347923e68dfb6de773/packages/vite/src/node/plugins/assetImportMetaUrl.ts#L78), which seems like an odd thing to handle, unless `self.location` is 1:1 to `import.meta.url`?

Or the other way round is we keep it and change this code to use `self.location` instead, and could go even further to change the whole url to `selft.location.pathname`. But nonetheless, I don't think this would affect the PR.

Reply:
Totally. I didn't do it in this PR because I felt it was a bit out of scope. But for v3, we should do it once modern means `import.meta.url` support. If you think we should already do it as part of this PR, we can keep working here

Reply:
I think we can do it later on in a separate PR. Just put something to think about 😬